src/algorithms/sdp.jl

```diff
@@ -0,0 +1,137 @@
+# Semidefinite programming
+
+# Decompress
+function lowertriangind(mat::Matrix)
+    indices = [i for i in CartesianIndices(mat) if i[1]>i[2]]
+    return LinearIndices(mat)[indices]
+end
+
+function rearrange_x(x_L::AbstractVector, x_D::AbstractVector)
+    mat_dim = length(x_D)
+    L = zeros(mat_dim, mat_dim)
+    L[lowertriangind(L)] .= x_L
+    D = diagm(x_D)
+    return (L, D)
+end
+
+function decompress_symmetric(L::Matrix, D::Matrix)
+    L + D + L'
+end
+
+function decompress_symmetric(x_L::AbstractArray, x_D::AbstractArray)
+    L, D = rearrange_x(x_L, x_D)
+    return decompress_symmetric(L, D)
+end
+
+function ChainRulesCore.rrule(::typeof(rearrange_x), x_L::AbstractVector, x_D::AbstractVector)
+    function pullback((ΔL, ΔD))
+        Δx_L = ΔL[lowertriangind(ΔL)]
+        Δx_D = diag(ΔD)
+        NoTangent(), Δx_L, Δx_D 
+    end
+    return rearrange_x(x_L, x_D), pullback
+end
+
+
+# Optimizer (algorithm)
+struct SDPBarrierAlg{Alg <: AbstractOptimizer} <: AbstractOptimizer
+    sub_alg::Alg
+end
+function SDPBarrierAlg(;sub_alg)
+    return SDPBarrierAlg(sub_alg)
+end
+
+# Options
+mutable struct SDPBarrierOptions
+    # Dimension of objective matrix
+    # Hyperparameters 
+    # Initial value of `c` in barrier method: 
+    # `Real` for using same value for all `sd_constraints`, `AbstractArray` for assign them respectively
+    c_init::Union{Real, AbstractArray}
+    # Decrease rate of `c` for every epoch, same as above
+    c_decr::Union{Real, AbstractArray}
+    n_iter::Int
+    # sub_option to solve (in)equality constraints
+    sub_options
+    # Keep all results or not
+    keep_all::Bool
+end
+function SDPBarrierOptions(c_init, c_decr, n_iter; sub_options, keep_all=true)
```

Comment:
Can we have some default values here? 1.0, 0.1 and 10 seem reasonable to me. Also let's make all of them keyword arguments.

Reply:
Oh I see they are kwargs in the other method. Ignore my second comment.